File: src/noise.cpp
#include "../include/geekatplay/noise.hpp"
#include <random>
#include <algorithm>
#include <numbers>

namespace Geekatplay {

std::vector<float> NoiseGenerator::Perlin2D(int res, float frequency, uint32_t seed) {
    std::vector<float> heightfield(res * res, 0.0f);
    int gridSize = std::max(static_cast<int>(std::ceil(frequency)) + 2, 4);

    std::mt19937 rng(seed);
    std::uniform_real_distribution<float> dist(0.0f, 2.0f * 3.1415926535f);

    // Generate random unit gradient vectors
    std::vector<float> gradX(gridSize * gridSize);
    std::vector<float> gradY(gridSize * gridSize);
    for (int i = 0; i < gridSize * gridSize; ++i) {
        float angle = dist(rng);
        gradX[i] = std::cos(angle);
        gradY[i] = std::sin(angle);
    }

    for (int y = 0; y < res; ++y) {
        for (int x = 0; x < res; ++x) {
            float gx = (static_cast<float>(x) / res) * frequency;
            float gy = (static_cast<float>(y) / res) * frequency;

            int x0 = static_cast<int>(std::floor(gx)) % gridSize;
            int y0 = static_cast<int>(std::floor(gy)) % gridSize;
            int x1 = (x0 + 1) % gridSize;
            int y1 = (y0 + 1) % gridSize;

            float fx = gx - std::floor(gx);
            float fy = gy - std::floor(gy);

            float u = Fade(fx);
            float v = Fade(fy);

            // Dot products
            float d00 = gradX[y0 * gridSize + x0] * fx + gradY[y0 * gridSize + x0] * fy;
            float d10 = gradX[y0 * gridSize + x1] * (fx - 1.0f) + gradY[y0 * gridSize + x1] * fy;
            float d01 = gradX[y1 * gridSize + x0] * fx + gradY[y1 * gridSize + x0] * (fy - 1.0f);
            float d11 = gradX[y1 * gridSize + x1] * (fx - 1.0f) + gradY[y1 * gridSize + x1] * (fy - 1.0f);

            float nx0 = d00 + u * (d10 - d00);
            float nx1 = d01 + u * (d11 - d01);
            float n = nx0 + v * (nx1 - nx0);

            // Normalize to [0, 1]
            heightfield[y * res + x] = std::clamp((n + 0.707f) / 1.414f, 0.0f, 1.0f);
        }
    }
    return heightfield;
}

std::vector<float> NoiseGenerator::RidgedMultifractal(
    int res, int octaves, float frequency, float lacunarity, float gain, float offset, uint32_t seed
) {
    std::vector<float> heightfield(res * res, 0.0f);
    std::vector<float> weight(res * res, 1.0f);
    float freq = frequency;
    float amp = 1.0f;

    for (int o = 0; o < octaves; ++o) {
        auto layer = Perlin2D(res, freq, seed + o * 137);
        for (int i = 0; i < res * res; ++i) {
            float signal = offset - std::abs(layer[i] * 2.0f - 1.0f);
            signal = signal * signal * weight[i];
            heightfield[i] += signal * amp;
            weight[i] = std::clamp(signal * gain, 0.0f, 1.0f);
        }
        freq *= lacunarity;
        amp *= gain;
    }

    float minVal = *std::min_element(heightfield.begin(), heightfield.end());
    float maxVal = *std::max_element(heightfield.begin(), heightfield.end());
    if (maxVal > minVal) {
        for (auto& v : heightfield) {
            v = (v - minVal) / (maxVal - minVal);
        }
    }
    return heightfield;
}

std::vector<float> NoiseGenerator::VoronoiCellular(int res, int cellCount, uint32_t seed) {
    std::vector<float> heightfield(res * res, 0.0f);
    std::mt19937 rng(seed);
    std::uniform_real_distribution<float> dist(0.0f, 1.0f);

    std::vector<float> ptsX(cellCount);
    std::vector<float> ptsY(cellCount);
    for (int i = 0; i < cellCount; ++i) {
        ptsX[i] = dist(rng);
        ptsY[i] = dist(rng);
    }

    for (int y = 0; y < res; ++y) {
        for (int x = 0; x < res; ++x) {
            float nx = static_cast<float>(x) / res;
            float ny = static_cast<float>(y) / res;
            float minDistSq = 1e9f;

            for (int i = 0; i < cellCount; ++i) {
                float dx = nx - ptsX[i];
                float dy = ny - ptsY[i];
                float d2 = dx * dx + dy * dy;
                if (d2 < minDistSq) minDistSq = d2;
            }
            heightfield[y * res + x] = std::sqrt(minDistSq);
        }
    }

    float minVal = *std::min_element(heightfield.begin(), heightfield.end());
    float maxVal = *std::max_element(heightfield.begin(), heightfield.end());
    if (maxVal > minVal) {
        for (auto& v : heightfield) {
            v = (v - minVal) / (maxVal - minVal);
        }
    }
    return heightfield;
}

std::vector<float> NoiseGenerator::Billow(int res, int octaves, float frequency, uint32_t seed) {
    std::vector<float> heightfield(res * res, 0.0f);
    float freq = frequency;
    float amp = 1.0f;
    float totalAmp = 0.0f;

    for (int o = 0; o < octaves; ++o) {
        auto layer = Perlin2D(res, freq, seed + o * 99);
        for (int i = 0; i < res * res; ++i) {
            float billow = 2.0f * std::abs(layer[i] - 0.5f);
            heightfield[i] += billow * amp;
        }
        totalAmp += amp;
        freq *= 2.0f;
        amp *= 0.5f;
    }

    for (auto& v : heightfield) {
        v /= (totalAmp > 0.0f ? totalAmp : 1.0f);
    }
    return heightfield;
}

} // namespace Geekatplay

File: core/noise.py
"""
Geekatplay Studio — High-Performance Procedural Noise Generators.
Implements vectorized Perlin, Simplex, Ridged Multifractal, Voronoi, and Strata heightfield kernels.
"""

import numpy as np
from typing import Tuple, Optional


def generate_grid_coords(res: int, scale: float = 1.0) -> Tuple[np.ndarray, np.ndarray]:
    """Generates normalized [0, 1] grid coordinates for heightfield synthesis."""
    lin = np.linspace(0.0, scale, res, endpoint=False, dtype=np.float32)
    return np.meshgrid(lin, lin, indexing="xy")


def _fade(t: np.ndarray) -> np.ndarray:
    """Ken Perlin's quintic polynomial s-curve: 6t^5 - 15t^4 + 10t^3."""
    return t * t * t * (t * (t * 6.0 - 15.0) + 10.0)


def perlin_noise_2d(res: int, frequency: float = 4.0, seed: int = 42) -> np.ndarray:
    """Generates smooth, deterministic 2D Perlin gradient noise."""
    rng = np.random.default_rng(seed)
    grid_size = max(int(np.ceil(frequency)) + 2, 4)
    
    # Random unit gradient vectors
    angles = rng.uniform(0, 2.0 * np.pi, size=(grid_size, grid_size)).astype(np.float32)
    gradients = np.stack([np.cos(angles), np.sin(angles)], axis=-1)

    x, y = generate_grid_coords(res, scale=frequency)
    x0 = np.floor(x).astype(int)
    y0 = np.floor(y).astype(int)
    x1 = (x0 + 1) % grid_size
    y1 = (y0 + 1) % grid_size
    x0 = x0 % grid_size
    y0 = y0 % grid_size

    fx = x - np.floor(x)
    fy = y - np.floor(y)

    u = _fade(fx)
    v = _fade(fy)

    # Dot products with 4 corner gradients
    g00 = gradients[y0, x0]
    g10 = gradients[y0, x1]
    g01 = gradients[y1, x0]
    g11 = gradients[y1, x1]

    d00 = g00[..., 0] * fx + g00[..., 1] * fy
    d10 = g10[..., 0] * (fx - 1.0) + g10[..., 1] * fy
    d01 = g01[..., 0] * fx + g01[..., 1] * (fy - 1.0)
    d11 = g11[..., 0] * (fx - 1.0) + g11[..., 1] * (fy - 1.0)

    # Bilinear interpolation with s-curves
    nx0 = d00 + u * (d10 - d00)
    nx1 = d01 + u * (d11 - d01)
    noise = nx0 + v * (nx1 - nx0)

    # Normalize to [0, 1]
    return np.clip((noise + 0.707) / 1.414, 0.0, 1.0).astype(np.float32)


def ridged_multifractal(
    res: int,
    octaves: int = 8,
    frequency: float = 2.0,
    lacunarity: float = 2.15,
    gain: float = 0.5,
    offset: float = 1.0,
    seed: int = 42
) -> np.ndarray:
    """
    Generates sharp, alpine mountain ridges using Ridged Multifractal synthesis.
    Ideal for steep granite mountain peaks, arêtes, and razor-edge ridgelines.
    """
    heightfield = np.zeros((res, res), dtype=np.float32)
    weight = np.ones((res, res), dtype=np.float32)
    freq = frequency
    amplitude = 1.0

    for i in range(octaves):
        layer = perlin_noise_2d(res, frequency=freq, seed=seed + i * 137)
        # Invert and square to create ridges
        signal = offset - np.abs(layer * 2.0 - 1.0)
        signal = signal * signal
        signal = signal * weight

        heightfield += signal * amplitude
        weight = np.clip(signal * gain, 0.0, 1.0)
        freq *= lacunarity
        amplitude *= gain

    min_val = np.min(heightfield)
    max_val = np.max(heightfield)
    if max_val > min_val:
        heightfield = (heightfield - min_val) / (max_val - min_val)
    return heightfield.astype(np.float32)


def voronoi_cellular_noise(
    res: int,
    cell_count: int = 16,
    seed: int = 42,
    mode: str = "f1"
) -> np.ndarray:
    """
    Generates Voronoi/Worley cellular distance fields.
    Useful for basalt column formations, cracked dry mud, crater clusters, and terrace cells.
    """
    rng = np.random.default_rng(seed)
    feature_points = rng.uniform(0.0, 1.0, size=(cell_count, 2)).astype(np.float32)

    x, y = generate_grid_coords(res, scale=1.0)
    grid_pts = np.stack([x, y], axis=-1)  # (res, res, 2)

    # Compute distances to all feature points
    # (res, res, 1, 2) - (1, 1, cell_count, 2)
    diff = grid_pts[:, :, np.newaxis, :] - feature_points[np.newaxis, np.newaxis, :, :]
    dist_sq = np.sum(diff * diff, axis=-1)  # (res, res, cell_count)

    sorted_dists = np.sort(np.sqrt(dist_sq), axis=-1)
    f1 = sorted_dists[..., 0]
    f2 = sorted_dists[..., 1] if cell_count > 1 else f1

    if mode == "f2_minus_f1":
        field = f2 - f1
    elif mode == "f2":
        field = f2
    else:
        field = f1

    min_val = np.min(field)
    max_val = np.max(field)
    if max_val > min_val:
        field = (field - min_val) / (max_val - min_val)
    return field.astype(np.float32)


def billow_noise(
    res: int,
    octaves: int = 6,
    frequency: float = 3.0,
    seed: int = 42
) -> np.ndarray:
    """Generates puffy billow formations suitable for rolling hills, sand dunes, and clouds."""
    heightfield = np.zeros((res, res), dtype=np.float32)
    freq = frequency
    amplitude = 1.0
    total_amp = 0.0

    for i in range(octaves):
        layer = perlin_noise_2d(res, frequency=freq, seed=seed + i * 99)
        billow_layer = 2.0 * np.abs(layer - 0.5)
        heightfield += billow_layer * amplitude
        total_amp += amplitude
        freq *= 2.0
        amplitude *= 0.5

    return (heightfield / total_amp).astype(np.float32)
